citesphere-model/src/main/java/edu/asu/diging/citesphere/data/bib/impl/CitationVersionDao.java

```diff
@@ -0,0 +1,76 @@
+package edu.asu.diging.citesphere.data.bib.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.javers.core.Javers;
+import org.javers.core.metamodel.object.CdoSnapshot;
+import org.javers.core.metamodel.object.SnapshotType;
+import org.javers.repository.jql.QueryBuilder;
+import org.javers.shadow.Shadow;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import edu.asu.diging.citesphere.data.bib.ICitationVersionsDao;
+import edu.asu.diging.citesphere.model.bib.ICitation;
+import edu.asu.diging.citesphere.model.bib.impl.Citation;
+import edu.asu.diging.citesphere.model.bib.impl.CitationVersion;
+
+@Repository
+public class CitationVersionDao implements ICitationVersionsDao {
+
+    @Autowired
+    private Javers javers;
+    
+    private static final String GROUP_PROPERTY = "group";
+
+    @Override
+    public List<CitationVersion> getVersions(String groupId, String key, int page, int pageSize) {
+        QueryBuilder jqlQuery = QueryBuilder.byInstanceId(key, Citation.class);
+        List<CdoSnapshot> versions = javers.findSnapshots(jqlQuery.build()).stream()
+                .filter(snapshot -> snapshot.getType() != SnapshotType.TERMINAL
+                        && groupId.equals((String) snapshot.getPropertyValue(GROUP_PROPERTY)))
+                .collect(Collectors.toList());
+        int offset = page * pageSize;
+        if (offset >= versions.size()) {
+            return new ArrayList<>();
+        }
+        List<CitationVersion> itemVersions = versions.subList(offset, Math.min(offset + pageSize, versions.size()))
```

Comment:
instead of loading everything into memory and then take a subset of the results, your query should only get the ones you really need.

Reply:
The problem here is that the javers query builder does not have "not" logical operator. So, I cannot write a query which filters away the snapshots of type "TERMINAL"(deleted). It also does not take multiple criteria (create, update) for the query. Hence, I have to manually filter out the relevant snapshots after fetching all of them. And only after that I can extract the correct page.

Reply:
What if you use `withSnapshotType​` or `withSnapshotType​Update` on the query? then you would only have to add the initial commit in case of the first page, wouldn't you?

Reply:
Yeah, it should work. I will try it out.